---
title: "R assignment"
author: "Sushma Ambekar"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## BCB546C R assignment

#### Installing and loading required libraries
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("reshape2")) install.packages("reshape2")
if (!require("tidyr")) install.packages("tidyr")
if (!require("plyr")) install.packages("plyr")
if (!require("dplyr")) install.packages("dplyr")
library(tidyverse)
library(ggplot2)
library(reshape2)
library(tidyr)
library(plyr)
library(dplyr)
```

#### Obtaining required files from the class github repository

To get the data from the respository to R studio, I used the following code. It is important here to mention that the files have a header and are tab delimited

```{r}
fang_et_al_genotypes <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
snp_position <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/master/assignments/UNIX_Assignment/snp_position.txt", header = TRUE, sep = "\t")
```

#### File inspection

str() outputs the internal structure of an R object, a diagnostic function and an alternative to summary()
dim() gives the lengths of rows and colummns 
colnames() displays the column names
head() along with select() is being used here to show selected columns and rows

```{r}
head(fang_et_al_genotypes, n = 5) %>% select(1:5)
str(fang_et_al_genotypes)
dim(fang_et_al_genotypes)
colnames(fang_et_al_genotypes)
table(fang_et_al_genotypes$Group)
```



```{r}
dim(snp_position)
str(snp_position)
colnames(snp_position)
head(snp_position, n = 5) %>% select(1:5)
table(snp_position$Chromosome)
```


To create separate files with required maize and teosinte IDs, I used the following codes:

```{r}
maize_IDs <- fang_et_al_genotypes[which(fang_et_al_genotypes$Group == "ZMMIL" | fang_et_al_genotypes$Group == "ZMMLR" | fang_et_al_genotypes$Group == "ZMMMR"),]
teosinte_IDs <- fang_et_al_genotypes[which(fang_et_al_genotypes$Group == "ZMPBA" | fang_et_al_genotypes$Group == "ZMPIL" | fang_et_al_genotypes$Group == "ZMPJA"),]
```

Since we require only three columns from the snp_position file, "SNP_ID","Chromosome" and "Position", I combined the values of 1,3 and 4 columns into a new file

```{r}
snp_position_filtered <- snp_position[,c(1,3,4)]
```

In order to put the required files together, I first transposed them and added the row name, SNP_ID which will be the identifier while merging. 

```{r}
transposed_maize_genotypes <- as.data.frame(t(maize_IDs[,-1]))
colnames(transposed_maize_genotypes) <- maize_IDs$Sample_ID
transposed_teosinte_genotypes <- as.data.frame(t(teosinte_IDs[,-1]))
colnames(transposed_teosinte_genotypes) <- teosinte_IDs$Sample_ID

transposed_maize_genotypes <- cbind(SNP_ID = row.names(transposed_maize_genotypes), transposed_maize_genotypes)
transposed_teosinte_genotypes <- cbind(SNP_ID = row.names(transposed_teosinte_genotypes), transposed_teosinte_genotypes)
```

Then, I merged the intermediate files by "SNP_ID" and row names maintaining the order, snp_position_filtered followed by maize and teosinte genotypes. 

```{r}
maize_all_chromosomes <- merge(snp_position_filtered,transposed_maize_genotypes, by = "SNP_ID")
teosinte_all_chromosomes <- merge(snp_position_filtered,transposed_teosinte_genotypes, by = "SNP_ID")
```

Missing data in R is replaced by NA by default. I could not figure out how to do this while merging, so used the approach below.

```{r}
maize_all_chromosomes[maize_all_chromosomes=="NA"] <- "?/?"
teosinte_all_chromosomes[teosinte_all_chromosomes=="NA"] <- "?/?"
```

To sort our merged files by increasing and decreasing position, we use the order() function.

```{r}
maize_all_chromosomes_increasing <- maize_all_chromosomes[order(maize_all_chromosomes$Position),]
teosinte_all_chromosomes_increasing <- teosinte_all_chromosomes[order(teosinte_all_chromosomes$Position),]
maize_all_chromosomes_decreasing <- maize_all_chromosomes[order(maize_all_chromosomes$Position,decreasing = T),]
teosinte_all_chromosomes_decreasing <- teosinte_all_chromosomes[order(teosinte_all_chromosomes$Position,decreasing = T),]

```

Then we use for loops to subset our datasets by chromosome and write to new files
```{r}
for (i in 1:10){
 merge_maize_chromosome <- maize_all_chromosomes_increasing[maize_all_chromosomes_increasing$Chromosome == i,]
  write.csv(merge_maize_chromosome, file=paste("maize_increasing_",i,".csv",sep=""))
}

for (i in 1:10){
  merge_maize_chromosome_rev <- maize_all_chromosomes_decreasing[maize_all_chromosomes_decreasing$Chromosome == i,]
  suppressWarnings(merge_maize_chromosome_rev[merge_maize_chromosome_rev == "?/?"] <- factor("-/-"))
  write.csv(merge_maize_chromosome_rev, file=paste("maize_decreasing_",i,".csv",sep=""))
}

for (i in 1:10){
  merge_teosinte_chromosome <- teosinte_all_chromosomes_increasing[teosinte_all_chromosomes_increasing$Chromosome == i,]
  write.csv(merge_teosinte_chromosome, file=paste("teosinte_increasing_",i,".csv",sep=""))
}

for (i in 1:10){
  merge_teosinte_chromosome_rev <- teosinte_all_chromosomes_decreasing[teosinte_all_chromosomes_decreasing$Chromosome == i,]
  suppressWarnings(merge_teosinte_chromosome_rev[merge_teosinte_chromosome_rev == "?/?"] <- factor("-/-"))
  write.csv(merge_teosinte_chromosome_rev, file=paste("teosinte_decreasing_",i,".csv",sep=""))
}
```

## Plots

#### SNPs per chromosome

```{r, echo=FALSE}
fang_transposed <- as.data.frame(t(fang_et_al_genotypes[,-1]))
fang_transposed <- cbind(SNP_ID = row.names(fang_transposed), fang_transposed)
merge_fang <- merge(snp_position, fang_transposed, by = "SNP_ID")
merge_fang$Chromosome <- factor(merge_fang$Chromosome,levels = c(1:10,"unknown","multiple"))
ggplot(merge_fang, aes(Chromosome)) + geom_bar(aes(), fill="#FDB863") + theme_classic() + labs(x="Chromosome",y="# of SNPs")
```

To see what groups contribute the most SNPs, we can do a simple plot with the original data.

```{r, echo=FALSE}
ggplot(fang_et_al_genotypes, aes(Group)) + geom_bar(aes(), fill="violetred4") + theme_classic() + theme(axis.text.x = element_text(angle = 90)) + labs(x="Group",y="# of SNPs")
```

To plot missing data and heterozygosity, we need to first process the files to show these data. We can create a new column that records whether the genotype characters match. In all of fang_et_al_genotypes, we can see a general proportion of homozygous to heterozygous to missing data. 

```{r, echo=FALSE}
fang_melted <- melt(fang_et_al_genotypes, id = c("Sample_ID", "Group"))
fang_melted[fang_melted == "?/?" ] = NA
fang_melted$Homozygous <- (fang_melted$value=="A/A" | fang_melted$value=="C/C" | fang_melted$value=="G/G" | fang_melted$value=="T/T")
ggplot(fang_melted, aes(Homozygous)) + geom_bar(aes(), fill="deepskyblue4", width = 0.4) + theme_classic()
```

To group this data by the species and group we will make some new intermediate data frames.
```{r, echo=FALSE}
fang_melted_group <- fang_melted[order(fang_melted$Group),]
fang_melted_ID <- fang_melted[order(fang_melted$Sample_ID),]
```


To get the proportion of homo/heterozygosity for each ordered dataframe, I summarized the 'Homozygous' column. 

```{r}
fang_melted_group_prop <- ddply(fang_melted_group, c("Group"), summarise, homo=sum(Homozygous, na.rm=TRUE), 
                     heter=sum(Homozygous, na.rm=FALSE), NotValid=sum(is.na(Homozygous)))
group_prop_melt <- melt(fang_melted_group_prop, measure.vars = c("homo", "heter", "NotValid"))
fang_melted_ID_prop <- ddply(fang_melted_ID, c("Sample_ID"), summarise, homo=sum(Homozygous, na.rm=TRUE), 
                     heter=sum(Homozygous, na.rm=FALSE), NotValid=sum(is.na(Homozygous)))
ID_prop_melt <- melt(fang_melted_ID_prop, measure.vars = c("homo", "heter", "NotValid"))
ggplot(group_prop_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack") + theme_classic() + theme(axis.text.x = element_text(angle = 90))
ggplot(ID_prop_melt,aes(x = Sample_ID, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack") + theme_classic()
```

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
